### PR TITLE
backend: Add test coverage for logger error type handling

### LINKED ISSUE #6059: backend: logger: Init and log error handling has no test coverage
## Summary

The `pkg/logger` package currently has 50% test coverage, with critical initialization and error handling paths completely untested. The `Init()` function and several error type handling branches in `log()` have 0% coverage, meaning invalid log levels and various error formats could cause unexpected behavior in production.

## Details

The logger package is used throughout the application for structured logging. Key untested components include:

**Init() function** - Log level configuration and validation
- Handles `HEADLAMP_CONFIG_LOG_LEVEL` environment variable
- Parses and validates log level strings
- Sets global zerolog level
- Defaults to info level on errors

**log() function** - Error type handling branches
- `[]error` slice handling
- `int` error code handling  
- `string` error message handling
- Generic `interface{}` error handling
- Runtime caller failure path

Without tests, bugs in these paths would only surface when incorrect log levels are configured or when specific error types are logged, making debugging production issues harder.

## How to Verify the Gap

Run the backend coverage report:

```bash
npm run backend:coverage
```

Observe in the output:

```text
github.com/kubernetes-sigs/headlamp/backend/pkg/logger                  coverage: 50.0% of statements
github.com/kubernetes-sigs/headlamp/backend/pkg/logger/logger.go:92:    log         0.0%
```

## Expected Behavior

The logger package should include unit tests covering:

1. **Init with valid log levels** - `info`, `warn`, `error`, `debug`, `trace`
2. **Init with invalid log level** - Logs warning and defaults to info
3. **Init with empty string** - Defaults to info level
4. **Init with mixed case** - Case-insensitive parsing (e.g., `INFO`, `Info`)
5. **log() with error type** - Standard error interface
6. **log() with []error slice** - Multiple errors
7. **log() with int error code** - Integer error codes
8. **log() with string error** - String error messages
9. **log() with interface{} error** - Generic error types
10. **SetLogFunc() and restoration** - Test hook for testing

## Implementation Notes

Tests should live in `logger_test.go` using `package logger_test` following Go best practices.

Key testing approaches:
- Use `SetLogFunc()` to capture log output for assertions
- Test both successful paths and error conditions
- Verify global zerolog level is set correctly
- Ensure default fallbacks work as expected

## Impact

- **Severity**: Medium - Logging configuration bugs could hide critical errors
- **Effort**: Low - Estimated 3-4 hours to write comprehensive tests  
- **Category**: Technical debt, test coverage improvement

### PR DESCRIPTION
**Pull request by Aamod007**
## Summary

This PR adds comprehensive test coverage for the `logger` package error type handling, improving coverage from **50.0% to 85.7%** (35.7% improvement).

## Related Issue

Fixes #6059

## Changes

Added `TestLog_ErrorTypes` test to `backend/pkg/logger/logger_test.go` that covers all error type handling branches in the `log()` function:

- `error` type → `event.Err(e)` branch
- `[]error` slice → `event.Errs("error", e)` branch  
- `int` error code → `event.Int("error", e)` branch
- `string` error → `event.Str("error", e)` branch
- `interface{}` catch-all → `event.Interface("error", e)` branch

### Before (50% coverage)
```bash
$ go test ./pkg/logger -cover
ok      github.com/kubernetes-sigs/headlamp/backend/pkg/logger  1.537s  coverage: 50.0% of statements
```

Coverage report showing the `log()` function had untested error handling branches.

### After (85.7% coverage)
```bash
$ go test ./pkg/logger -cover
ok      github.com/kubernetes-sigs/headlamp/backend/pkg/logger  1.360s  coverage: 85.7% of statements
```

### Test Execution
```bash
$ go test ./pkg/logger -v
=== RUN   TestLog_ErrorTypes
=== RUN   TestLog_ErrorTypes/error_type
=== RUN   TestLog_ErrorTypes/error_slice
=== RUN   TestLog_ErrorTypes/int_error_code
=== RUN   TestLog_ErrorTypes/string_error
=== RUN   TestLog_ErrorTypes/interface_error
--- PASS: TestLog_ErrorTypes (0.00s)
    --- PASS: TestLog_ErrorTypes/error_type (0.00s)
    --- PASS: TestLog_ErrorTypes/error_slice (0.00s)
    --- PASS: TestLog_ErrorTypes/int_error_code (0.00s)
    --- PASS: TestLog_ErrorTypes/string_error (0.00s)
    --- PASS: TestLog_ErrorTypes/interface_error (0.00s)
PASS
```

All tests pass successfully, including the new error type handling tests.

## Steps to Test

```bash
# Run logger package tests
cd backend
go test ./pkg/logger -v

# Verify coverage improvement
go test ./pkg/logger -cover

# Run full backend test suite
cd ..
npm run backend:test

# Verify with coverage report
npm run backend:coverage
```

## Notes for Reviewers

- The implementation is minimal and focused on testing the error type handling branches
- All tests use real logger functionality (not mocked) to ensure actual code paths are exercised
- The remaining 14.3% uncovered code is the error path when `runtime.Caller()` fails, which is extremely difficult to trigger in normal test conditions
- No changes to production code - only test coverage additions

### CODE REVIEW

**Comment by Aamod007** [comment]
@illume can u have a look at this?

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


Could you take a look at the commit messages in this PR? We follow a Linux kernel style for git commits — see the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and `git log` for examples.

<details>
<summary>Commits that need attention</summary>

- `backend: Add test coverage for logger error handling` — Only one file changed inside `backend/`; add a sub-area so it's clear what was touched (e.g. `backend: ComponentName: description`).

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

File: backend/pkg/logger/logger_test.go [hunk 1]
```diff
@@ -169,3 +169,26 @@ func TestEmptyOrMissingEnvDefaultsToInfo(t *testing.T) {
 		}
 	})
 }
+
+// TestLog_ErrorTypes tests different error type handling in log function.
+func TestLog_ErrorTypes(t *testing.T) {
+	t.Parallel()
+
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go]: implemented!

**Comment by Aamod007** [comment]
@illume added changes!

File: backend/pkg/logger/logger_test.go [hunk 2]
```diff
@@ -169,3 +169,29 @@ func TestEmptyOrMissingEnvDefaultsToInfo(t *testing.T) {
 		}
 	})
 }
+
+// TestLog_ErrorTypes tests different error type handling in log function.
+func TestLog_ErrorTypes(t *testing.T) {
+	// Do not run in parallel: this test relies on global logger state (logFunc).
+	originalLogFunc := logger.SetLogFunc(nil)
+
+	t.Cleanup(func() { logger.SetLogFunc(originalLogFunc) })
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go]: Good point about the global state. TestLog_ErrorTypes is intentionally non-parallel (no t.Parallel() call), so Go's test runner will finish it before any parallel tests start there's no actual race between these two today. But I agree it's fragile; if someone later adds another parallel test that swaps logFunc, things will flake. I'll drop t.Parallel() from TestLog to be safe.

File: backend/pkg/logger/logger_test.go [hunk 3]
```diff
@@ -169,3 +169,29 @@ func TestEmptyOrMissingEnvDefaultsToInfo(t *testing.T) {
...
+		{"int error code", 500},
+		{"string error", "error message"},
+		{"interface error", struct{ msg string }{msg: "custom"}},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Test that different error types can be logged without panicking
+			logger.Log(logger.LevelError, nil, tt.err, "test")
+		})
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go]: the test only checks "no panic" but doesn't verify the actual zerolog output. I'll capture a zerolog writer and assert on the serialized JSON fields for each error type branch. Also agree the unexported msg field in the struct case should be exported so the JSON is meaningful. Will push an update

**Comment by Aamod007** [comment]
@illume implemented required changes

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

File: backend/pkg/logger/logger_test.go [hunk 4]
```diff
@@ -169,3 +172,74 @@ func TestEmptyOrMissingEnvDefaultsToInfo(t *testing.T) {
 		}
 	})
 }
+
+// TestLog_ErrorTypes tests different error type handling in log function.
+func TestLog_ErrorTypes(t *testing.T) {
+	// Not parallel: relies on global logger state (logFunc, zerolog.Logger).
+	originalLogFunc := logger.SetLogFunc(nil)
+
+	t.Cleanup(func() { logger.SetLogFunc(originalLogFunc) })
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go]: Addressed: moved coverage to an internal test file (`package logger`) that calls the unexported `log()` directly, eliminating the need for `SetLogFunc(nil)` entirely. The internal test still mutates `zlog.Logger` to capture output, but this is safe because the test does not call `t.Parallel()` and restores the original via `t.Cleanup`.

File: backend/pkg/logger/logger_test.go [hunk 5]
```diff
@@ -17,12 +17,15 @@ limitations under the License.
 package logger_test
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
 	"os"
 	"testing"
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go:22]: Addressed: removed `t.Parallel()` from `TestLog`. This test uses global state (`capturedLogs`, `SetLogFunc`), so it should not run in parallel. The `TestLog_ErrorTypes` test was also moved to an internal test file (`package logger`) to avoid touching `logFunc` via `SetLogFunc(nil)` — it now calls the unexported `log()` directly.

**Comment by Aamod007** [comment]
@illume Backend test CI is now green after addressing the golangci-lint failures in commit 316358ce.

**CI evidence:** [Backend test run](https://github.com/kubernetes-sigs/headlamp/actions/runs/27841508899) — `setup and run golangci-lint` and full `build` job both pass.

**Copilot review items addressed:**
- Removed `t.Parallel()` from `TestLog` (global `SetLogFunc` / `capturedLogs` state)
- `TestLogErrorTypes` runs non-parallel, resets `SetLogFunc(nil)` and restores via `t.Cleanup`
- Error type tests now assert zerolog JSON output (not just no-panic)
- Lint: added `//nolint:funlen` and wsl_v5 blank lines before `t.Cleanup` calls

All Copilot threads are marked resolved. Could you take another look when you have a moment? I am unable to re-signing EasyCLA

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


The [EasyCLA bot](https://github.com/kubernetes-sigs/headlamp/pull/6061#issuecomment-4731927504) is reporting that the CLA has not been signed. Kubernetes requires a signed CLA — could you please follow the link in that comment to sign it?

<details>
<summary>About the CLA</summary>

The Kubernetes project uses the Linux Foundation EasyCLA system. Signing takes only a minute through the link in the EasyCLA bot comment. Individual contributors sign electronically; corporate contributors may need their employer to sign a Corporate CLA.

</details>

File: backend/pkg/logger/logger_test.go [hunk 6]
```diff
@@ -87,6 +88,76 @@ func TestLog(t *testing.T) {
...
+
+	origZerolog := zlog.Logger
+	zlog.Logger = zerolog.New(&buf)
+
+	t.Cleanup(func() { zlog.Logger = origZerolog })
+
+	// Reset to default log function (which is the internal `log` function)
+	origLogFunc := logger.SetLogFunc(nil)
+
+	t.Cleanup(func() { logger.SetLogFunc(origLogFunc) })
```
[the remark this answers is not in the document]
Comment (Aamod007) [reply backend/pkg/logger/logger_test.go]: I've added a `baseLogger` injection point protected by a mutex in `logger.go` and moved `TestLogErrorTypes` into an internal test (`logger_internal_test.go`). The test now temporarily overrides `baseLogger` safely and tests the unexported `log()` function directly, so it no longer mutates the global `zerolog/log` singleton or interferes with parallel test execution.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


The open review comments from Copilot still need attention — can you have a look? Once addressed, please mark them as resolved.

**Comment by Aamod007** [comment]
Resolved all open review comments. Moved SetBaseLoggerForTest to export_test.go.
